File: typescript/src/days/day01.ts
const LONGDIGITS = '0 1 2 3 4 5 6 7 8 9 zero one two three four five six seven eight nine'.split(' ');
const DIGITS = '0 1 2 3 4 5 6 7 8 9'.split(' ')

export function day01(data: string) {
    const lines = data.split('\n');

    const value1 = sumFirstAndLast(lines, DIGITS);
    const value2 = sumFirstAndLast(lines, LONGDIGITS);

    console.log('The calibration value is', value1);
    console.log('The correct calibration value is', value2);
}

function sumFirstAndLast(lines: string[], digits: string[]): number {
    return lines.
        map(line => firstDigit(line, digits) * 10 + lastDigit(line, digits)).
        reduce((a, b) => a + b);
}

function firstDigit(s: string, digits: string[]): number {
    for (let p = 0; p <= s.length; p++) {
        for(let d = 0; d < digits.length; d++) {
            if (s.slice(0, p).endsWith(digits[d])) return d % 10;
        }
    }
    return 0;
}

function lastDigit(s: string, digits: string[]): number {
    for (let p = s.length-1; p >= 0; p--) {
        for(let d = 0; d < digits.length; d++) {
            if (s.slice(p).startsWith(digits[d])) return d % 10;
        }
    }
    return 0;
}

File: python/days/day01.py
"""
--- Day 01: Trebuchet Repairs ---
"""

_DIGITS = list("0123456789")
_NUMS = "zero one two three four five six seven eight nine".split()
_ALL = _DIGITS + _NUMS
_VALUES = dict(zip(_NUMS, _DIGITS))


def day01():
    """
    --- Day 01: main  ---
    """
    lines = read_input()
    print("--- Day 01: Trebuchet Repairs ---")
    print("Part 1:", part1(lines))
    print("Part 2:", part2(lines))


def part1(lines: list[str]) -> int:
    """
    --- Day 01: Part 1 ---
    """
    total = 0
    for line in lines:
        first, last = find_first_and_last(line, _DIGITS)
        total += int(first + last)
    return total


def part2(lines: list[str]) -> int:
    """
    --- Day 01: Part 2 ---
    """
    total = 0
    for line in lines:
        first, last = find_first_and_last(line, _ALL)
        first = _VALUES.get(first, first)
        last = _VALUES.get(last, last)
        total += int(first + last)
    return total


def find_first_and_last(line: str, digits: list[str]) -> tuple[str, str]:
    """
    Find the first and last digit in the line
    """
    first = ""
    first_pos = 1e6
    last = ""
    last_pos = -1e6
    for digit in digits:
        pos = line.find(digit)
        if pos != -1 and pos < first_pos:
            first = digit
            first_pos = pos
        pos = line.rfind(digit)
        if pos != -1 and pos > last_pos:
            last = digit
            last_pos = pos
    return first, last


def read_input():
    """
    Read the input file and return a list of lines
    """
    with open("../data/day01.txt", encoding="utf-8") as f:
        return f.readlines()
